## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'], axis=1)
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [3]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
counts = application_df['APPLICATION_TYPE'].value_counts()
counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(counts[counts < 500].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts for binning
class_counts = application_df['CLASSIFICATION'].value_counts()
class_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
class_counts_1 = class_counts.loc[class_counts > 1]
class_counts_1.head()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list(class_counts[class_counts < 1000].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()


C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [9]:
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [10]:
# Look at ASK_AMT value counts for binning
asks = application_df['ASK_AMT'].value_counts()
asks

5000        25398
10478           3
15583           3
63981           3
6725            3
            ...  
5371754         1
30060           1
43091152        1
18683           1
36500179        1
Name: ASK_AMT, Length: 8747, dtype: int64

In [11]:
# Choose a cutoff value and create a list of asks to be replaced
asks_to_replace = list(asks[asks < 6000].index)

# Replace in dataframe
for ask in asks_to_replace:
    application_df['ASK_AMT'] = application_df['ASK_AMT'].replace(
        ask, "Other")

# Check to make sure binning was successful
application_df['ASK_AMT'].value_counts()

5000     25398
Other     8901
Name: ASK_AMT, dtype: int64

In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
numeric = pd.get_dummies(application_df)
numeric

STATUS  IS_SUCCESSFUL  APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  \
0           1              1                       0                     1   
1           1              1                       0                     0   
2           1              0                       0                     0   
3           1              1                       0                     0   
4           1              1                       0                     0   
...       ...            ...                     ...                   ...   
34294       1              0                       0                     0   
34295       1              0                       0                     0   
34296       1              0                       0                     0   
34297       1              1                       0                     0   
34298       1              0                       0                     0   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                         0                    0                    0   
1                         0                    1                    0   
2                         0                    0                    0   
3                         0                    1                    0   
4                         0                    1                    0   
...                     ...                  ...                  ...   
34294                     0                    0                    1   
34295                     0                    0                    1   
34296                     0                    1                    0   
34297                     0                    0                    0   
34298                     0                    1                    0   

       APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  ...  \
0                        0                    0                    0  ...   
1                        0                    0                    0  ...   
2                        1                    0                    0  ...   
3                        0                    0                    0  ...   
4                        0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    0                    0                    0  ...   
34295                    0                    0                    0  ...   
34296                    0                    0                    0  ...   
34297                    1                    0                    0  ...   
34298                    0                    0                    0  ...   

       INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                             0                   0                 0   
1                             0                   0                 0   
2                             0                   0                 0   
3                             0                   0                 0   
4                             1                   0                 0   
...                         ...                 ...               ...   
34294                         0                   0                 0   
34295                         0                   0                 0   
34296                         0                   0                 0   
34297                         0                   0                 0   
34298                         0                   0                 1   

       INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  \
0                           0                0                  0   
1                           0                0                  0   
2                           0                0                  0   
3                           0                0                  0   
4                           0                0

In [13]:
# Split our preprocessed data into our features and target arrays
X = numeric.drop(["IS_SUCCESSFUL"], axis=1)
y = numeric["IS_SUCCESSFUL"]

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split (X, y, random_state=1)

In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [28]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
num_input_features = len(X_train_scaled[0])
layer_1 = 10
layer_2 = 8
layer_3 = 6

nn_mod2 = tf.keras.models.Sequential()

# First hidden layer
nn_mod2.add(tf.keras.layers.Dense(units=layer_1,
             input_dim=num_input_features, activation="relu"))

# Second hidden layer
nn_mod2.add(tf.keras.layers.Dense(units=layer_2,
             input_dim=num_input_features, activation="sigmoid"))

# Third hidden layer
nn_mod2.add(tf.keras.layers.Dense(units=layer_3,
             input_dim=num_input_features, activation="sigmoid"))

# Output layer
nn_mod2.add(tf.keras.layers.Dense(units=1,
             input_dim=num_input_features, activation="sigmoid"))

# Check the structure of the model
nn_mod2.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 10)                450       
                                                                 
 dense_13 (Dense)            (None, 8)                 88        
                                                                 
 dense_14 (Dense)            (None, 6)                 54        
                                                                 
 dense_15 (Dense)            (None, 1)                 7         
                                                                 
Total params: 599
Trainable params: 599
Non-trainable params: 0
_________________________________________________________________


In [29]:
# Compile the model
nn_mod2.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [30]:
# Train the model
fit_mod2 = nn_mod2.fit(X_train_scaled, y_train, epochs=20)

Epoch 1/20
804/804 [==============================] - 3s 2ms/step - loss: 0.6148 - accuracy: 0.6858
Epoch 2/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5664 - accuracy: 0.7277
Epoch 3/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5604 - accuracy: 0.7294
Epoch 4/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5576 - accuracy: 0.7302
Epoch 5/20
804/804 [==============================] - 2s 3ms/step - loss: 0.5557 - accuracy: 0.7304
Epoch 6/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5545 - accuracy: 0.7299
Epoch 7/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5532 - accuracy: 0.7301
Epoch 8/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5525 - accuracy: 0.7310
Epoch 9/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5517 - accuracy: 0.7311
Epoch 10/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5513 - accuracy: 0.7308

In [31]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_mod2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5560 - accuracy: 0.7300 - 512ms/epoch - 2ms/step
Loss: 0.5560420751571655, Accuracy: 0.7300291657447815


In [32]:
# Export our model to HDF5 file
nn_mod2.save('Alpha_Optimization_Attempt1.h5')


<b>Optimization Attempt #1</b>

APPLICATION_TYPE cutoff = 500

CLASSIFICATION cutoff = 1,000

Loss: 0.5530775189399719, Accuracy: 0.7313119769096375

layer1: 10; activation = relu

layer2: 8; activation = sigmoid

layer3: 6; activation = sigmoid

A loss value of 55% indicates that the model can be further optimized.

The accuracy percent shows that 73% of the model's predicted values align with the true values in the original dataset.

Additional attempts could be made to try to reach 75% accuracy.